File: Libraries/python/beagle_helper/src/beagle_helper/pin.py
import gpiod

from .abstractions.pwm import PwmChip, PwmChannel
from time import sleep


class GpioPin:
    def __init__(self, chipset: str, offset: int) -> None:
        self.chipset = chipset
        self.offset = offset

    def gpio_input(self):
        chip = gpiod.Chip(self.chipset)
        led = chip.get_line(self.offset)
        led.request(consumer="", type=gpiod.LINE_REQ_EV_BOTH_EDGES)

        return led


class PwmPin:
    def __init__(self, chipset: int, channel: int) -> None:
        self.chipset = chipset
        self.channel = channel

    def pwm_output(self) -> PwmChannel:
        chip = PwmChip(self.chipset)
        return chip.export(self.channel)


class Pin:
    def __init__(self, gpio: GpioPin | None = None, pwm: PwmPin | None = None) -> None:
        self.gpio = gpio
        self.pwm = pwm

    def gpio_input(self):
        if self.gpio == None:
            raise ValueError("Pin cannot be used as GPIO")

        return self.gpio.gpio_input()

    def pwm_output(self) -> PwmChannel:
        if self.pwm == None:
            raise ValueError("Pin cannot be used as PWM")

        chan = self.pwm.pwm_output()
        
        # Linux group permissions need a bit to work on the newly exported pwm.
        # Removing this delay can lead to Permission denied.
        sleep(0.1)

        return chan

File: Libraries/rust/beagle_helper/src/pin.rs
#![allow(dead_code)]

pub(crate) struct GpioPin {
    chip: usize,
    offset: u32,
}

impl GpioPin {
    pub(crate) const fn new(chip: usize, offset: u32) -> Self {
        Self { chip, offset }
    }
}

pub(crate) struct PwmPin {
    chip: usize,
    channel: usize,
}

impl PwmPin {
    pub(crate) const fn new(chip: usize, channel: usize) -> Self {
        Self { chip, channel }
    }

    #[cfg(feature = "pwm")]
    pub(crate) fn pwm_output(&self) -> std::io::Result<crate::abstractions::pwm::PwmChannel> {
        let chip = crate::abstractions::pwm::PwmChip::new(self.chip)?;
        let chan = chip.export(self.channel)?;

        // Linux group permissions need a bit to work on the newly exported pwm. Removing this delay
        // can lead to Permission denied.
        std::thread::sleep(std::time::Duration::from_millis(100));

        Ok(chan)
    }
}

pub(crate) struct AdcPin {
    iio_device: usize,
    channel: usize,
}

impl AdcPin {
    pub(crate) const fn new(iio_device: usize, channel: usize) -> Self {
        Self {
            iio_device,
            channel,
        }
    }
}

pub struct Pin {
    gpio: Option<GpioPin>,
    pwm: Option<PwmPin>,
}

impl Pin {
    pub(crate) const fn new(gpio: Option<GpioPin>, pwm: Option<PwmPin>) -> Self {
        Self { gpio, pwm }
    }

    #[cfg(feature = "pwm")]
    pub(crate) fn pwm_output(&self) -> std::io::Result<crate::abstractions::pwm::PwmChannel> {
        match &self.pwm {
            Some(x) => x.pwm_output(),
            None => Err(std::io::Error::other(
                "Pin does not support PWM output functionality",
            )),
        }
    }
}
